### PR TITLE
@0xproject/contract-wrappers V2 refactor. Part 2

### PR DESCRIPTION
**Pull request by LogvinovLeon**
## Description

This PR - migrates exchange contract wrappers and fill-scenarios to v2.
For now - it doesn't have validation except assertions and revert reasons.

## Testing instructions

Run the exchange wrapper tests in contract-wrappers

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [x] New feature (non-breaking change which adds functionality)
*   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [x] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [x] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [x] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 1]
```diff
@@ -0,0 +1,994 @@
...
+            {
+                from: normalizedTakerAddress,
+                gas: orderTransactionOpts.gasLimit,
+                gasPrice: orderTransactionOpts.gasPrice,
+            },
+        );
+        return txHash;
+    }
+    /**
+     * Synchronously executes multiple calls of fillOrder until total amount of makerAsset is bought by taker.
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: `of fillOrder` -> `to fillOrder`

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,994 @@
...
+     * @param   signedOrders                                    An array of signed orders to fill.
+     * @param   takerAssetFillAmount                            Taker asset fill amount.
+     * @param   takerAddress                                    The user Ethereum address who would like to fill
+     *                                                          these orders. Must be available via the supplied
+     *                                                          Provider passed to 0x.js.
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: `instantiation`

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,994 @@
...
+    }
+    /**
+     * Batch version of cancelOrderAsync. Executes multiple cancels atomically in a single transaction.
+     * @param   orders                                          An array of orders to cancel.
+     * @param   orderTransactionOpts                            Optional arguments this method accepts.
+     * @return  Transaction hash.
+     */
+    @decorators.asyncZeroExErrorHandler
+    public async batchCancelOrdersAsync(
+        orders: Order[],
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: Can they pass in `signedOrder`'s? If so, let's include that.

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 4]
```diff
@@ -0,0 +1,994 @@
...
+            gas: orderTransactionOpts.gasLimit,
+            gasPrice: orderTransactionOpts.gasPrice,
+        });
+        return txHash;
+    }
+    /**
+     * Match two complementary orders that have a profitable spread.
+     * Each order is filled at their respective price point. However, the calculations are carried out as though
+     * the orders are both being filled at the right order's price point.
+     * The profit made by the left order goes to the taker (who matched the two orders).
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: `whoever matched...`

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 5]
```diff
@@ -0,0 +1,994 @@
...
+        orderTransactionOpts: OrderTransactionOpts = {},
+    ): Promise<string> {
+        assert.doesConformToSchema('leftSignedOrder', leftSignedOrder, schemas.signedOrderSchema);
+        assert.doesConformToSchema('rightSignedOrder', rightSignedOrder, schemas.signedOrderSchema);
+        await assert.isSenderAddressAsync('takerAddress', takerAddress, this._web3Wrapper);
+        assert.doesConformToSchema('orderTransactionOpts', orderTransactionOpts, orderTxOptsSchema, [txOptsSchema]);
+        const normalizedTakerAddress = takerAddress.toLowerCase();
+        // TODO(logvinov): Check that:
+        // rightOrder.makerAssetData === leftOrder.takerAssetData;
+        // rightOrder.takerAssetData === leftOrder.makerAssetData;
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: When will this TODO be addressed?

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,994 @@
...
+    @decorators.asyncZeroExErrorHandler
+    public async isValidSignatureAsync(
+        hash: string,
+        signerAddress: string,
+        signature: string,
+        methodOpts: MethodOpts = {},
+    ): Promise<boolean> {
+        assert.isHexString('hash', hash);
+        assert.isETHAddressHex('signerAddress', signerAddress);
+        if (!_.isUndefined(methodOpts)) {
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: Let's add basic `is hex string` assertion for `signature`.

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 7]
```diff
@@ -0,0 +1,994 @@
...
+            hash,
+            signerAddress,
+            signature,
+            txData,
+            methodOpts.defaultBlock,
+        );
+        return isValidSignature;
+    }
+    /**
+     * Checks if the is allowed by the signer.
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: Keck?

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,994 @@
...
+        const isValidSignature = await exchangeInstance.allowedValidators.callAsync(
+            normalizedSignerAddress,
+            normalizedValidatorAddress,
+            txData,
+            methodOpts.defaultBlock,
+        );
+        return isValidSignature;
+    }
+    /**
+     * Approves a hash on-chain using any valid signature type.
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: Incorrect comment for method. Please fix.

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,994 @@
...
+        const contractAddress = this._getContractAddress(artifacts.ZRXToken, this._zrxContractAddressIfExists);
+        return contractAddress;
+    }
+    /**
+     * Returns the ZRX asset data used by the exchange contract.
+     * @return ZRX asset data
+     */
+    public async getZRXAssetDataAsync(): Promise<string> {
+        const exchangeInstance = await this._getExchangeContractAsync();
+        const ZRX_ASSET_DATA = exchangeInstance.ZRX_ASSET_DATA.callAsync();
```
Comment (fabioberger) [inline packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: Let's lowercase, camelcase this variable name.

File: packages/contract-wrappers/test/exchange_wrapper_test.ts [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+import { chaiSetup } from './utils/chai_setup';
+import { constants } from './utils/constants';
+import { tokenUtils } from './utils/token_utils';
+import { provider, web3Wrapper } from './utils/web3_wrapper';
+
+chaiSetup.configure();
+const expect = chai.expect;
+const blockchainLifecycle = new BlockchainLifecycle(web3Wrapper);
+
+// TODO(logvinov): Check coverage
```
Comment (fabioberger) [inline packages/contract-wrappers/test/exchange_wrapper_test.ts]: When will this be done?

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts (hunk 3 above)
Comment (LogvinovLeon) [reply packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: For docs generation reasons? OK

File: packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts (hunk 5 above)
Comment (LogvinovLeon) [reply packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: In the new PR. Part 3. I'll add missing validations there. This one doesn't have any validations and I'm not sure if it makes sense to add just one.

Comment (LogvinovLeon) [reply packages/contract-wrappers/src/contract_wrappers/exchange_wrapper.ts]: There will be also Part 4 that will assign those variables to NULL_ADDRESS to save gas.

**Pull request merged**